massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,642 @@
+    }
+
+    /// Check if it is a Denunciation of several block headers
+    fn is_for_block_header(&self) -> bool {
+        matches!(self, Denunciation::BlockHeader(_))
+    }
+
+    /// Check if it is a Denunciation for this endorsement
+    #[cfg(test)]
+    fn is_also_for_endorsement(
+        &self,
+        s_endorsement: &SecureShareEndorsement,
+    ) -> Result<bool, DenunciationError> {
+        match self {
+            Denunciation::BlockHeader(_) => Ok(false),
+            Denunciation::Endorsement(endo_de) => {
+                let content_hash =
+                    EndorsementDenunciation::compute_content_hash(&s_endorsement.content)?;
+
+                let hash_ = EndorsementDenunciation::compute_hash_for_sig_verif(
+                    &endo_de.public_key,
+                    &endo_de.slot,
+                    &endo_de.index,
+                    &content_hash,
+                );
+
+                if let Ok(hash) = hash_ {
+                    Ok(endo_de.slot == s_endorsement.content.slot
+                        && endo_de.index == s_endorsement.content.index
+                        && endo_de.public_key == s_endorsement.content_creator_pub_key
+                        && endo_de.hash_1 != content_hash
+                        && endo_de.hash_2 != content_hash
+                        && endo_de
+                            .public_key
+                            .verify_signature(&hash, &s_endorsement.signature)
+                            .is_ok())
+                } else {
+                    Ok(false)
+                }
+            }
+        }
+    }
+
+    /// Check if it is a Denunciation for this block header
+    #[cfg(test)]
+    fn is_also_for_block_header(
+        &self,
+        s_block_header: &SecuredHeader,
+    ) -> Result<bool, DenunciationError> {
+        match self {
+            Denunciation::Endorsement(_) => Ok(false),
+            Denunciation::BlockHeader(endo_bh) => {
+                let content_hash =
+                    BlockHeaderDenunciation::compute_content_hash(&s_block_header.content)?;
+
+                let hash_ = BlockHeaderDenunciation::compute_hash_for_sig_verif(
+                    &endo_bh.public_key,
+                    &endo_bh.slot,
+                    &content_hash,
+                );
+
+                if let Ok(hash) = hash_ {
+                    Ok(endo_bh.slot == s_block_header.content.slot
+                        && endo_bh.public_key == s_block_header.content_creator_pub_key
+                        && endo_bh.hash_1 != content_hash
+                        && endo_bh.hash_2 != content_hash
+                        && endo_bh
+                            .public_key
+                            .verify_signature(&hash, &s_block_header.signature)
+                            .is_ok())
+                } else {
+                    Ok(false)
+                }
+            }
+        }
+    }
+
+    fn is_valid(&self) -> Result<bool, DenunciationError> {
```

Comment:
is this also only for tests ?

Reply:
Add a TODO so this will be done in a follow up